### PR TITLE
core: process block announce message

### LINKED ISSUE #346: core: request block details after receiving BlockAnnounce message
A BlockAnnounce message only contains the block header, however to validate the block, we need the entire thing. In core.Service, we should create a BlockRequest for the full block after recieving a BlockAnnounce message.

note, BlockReponse is already handled so that should be ok. However we would like to have some sort of check that the BlockResponse we're getting matches up to our request.

### PR DESCRIPTION
**Pull request by ryanchristo**
### Changes

This pull request adds functionality to announce blocks from the BABE session and to process block announce messages from the p2p service.

We needed to create a new channel in order to send messages to the p2p service, which means we had to rename the channels.

## Tests:
```
go test ./core -v
```

### Issues:
closes #346

### CODE REVIEW

File: core/service.go [hunk 1]
```diff
@@ -28,117 +32,200 @@ import (
...
+		msg, ok := <-s.p2pRec
+		if !ok {
+			log.Debug("receive channel closed")
+			return // exit
+		}
+		err := s.handleMessage(msg)
 		if err != nil {
-			log.Error("ProcessTransaction", "error", err)
-			return err
+			log.Debug("handle message", "err", err)
```
Comment (ansermino) [inline core/service.go]: just listing out the function thats called, please update this based on our discussions on your other PR.

Should also be error.

File: core/service.go [hunk 2]
```diff
@@ -28,117 +32,200 @@ import (
...
+	// get transactions from message extrinsics
+	txs := msg.(*p2p.TransactionMessage).Extrinsics
+
+	for _, tx := range txs {
+		tx := tx // pin
+
+		// validate each transaction
+		val, err := s.validateTransaction(tx)
+		if err != nil {
+			log.Debug("validate transaction", "err", err)
```
Comment (ansermino) [inline core/service.go]: error == error <3

File: core/service.go [hunk 3]
```diff
@@ -28,117 +32,200 @@ import (
...
+// announce messages (block announce messages include the header but the full
+// block is required to execute `core_execute_block`).
+func (s *Service) ProcessBlockAnnounceMessage(msg p2p.Message) error {
+
+	// TODO: check to send block request message
+
+	encMsg, err := msg.Encode()
+	if err != nil {
+		return err
+	}
```
Comment (ansermino) [inline core/service.go]: Why encode it?

Comment (ryanchristo) [reply core/service.go]: removed

File: core/service.go (hunk 2 above)
Comment (ryanchristo) [reply core/service.go]: Fixed.

File: core/service.go (hunk 1 above)
Comment (ryanchristo) [reply core/service.go]: Fixed.

File: consensus/babe/babe.go [hunk 4]
```diff
@@ -205,10 +195,10 @@ func (b *Session) vrfSign(input []byte) ([]byte, error) {
...
 
 // BuildBlock Builds the block
-func (b *Session) buildBlock(number *big.Int) (*types.Block, error) {
+func (b *Session) buildBlock(number *big.Int) (types.Block, error) {
 	block := types.Block{
 		Header: types.BlockHeaderWithHash{Number: number},
 		Body:   []byte{1, 2, 3, 4, 5},
 	}
-	return &block, nil
+	return block, nil
```
Comment (noot) [inline consensus/babe/babe.go]: why is this not a ptr anymore?

File: core/service.go [hunk 5]
```diff
@@ -28,117 +32,195 @@ import (
...
+func NewService(rt *runtime.Runtime, bsRec chan types.Block, p2pRec <-chan p2p.Message, p2pSend chan<- p2p.Message) (*Service, error) {
+
+	// create a new BABE session (bsRec becomes blockSend in babe)
+	bs, err := babe.NewSession([32]byte{}, [64]byte{}, rt, bsRec)
 	if err != nil {
 		return nil, err
 	}
 
-	return &Service{
+	s := &Service{
```
Comment (noot) [inline core/service.go]: if we are just returning a struct it's ok to put it all in the return line

File: core/service.go [hunk 6]
```diff
@@ -28,117 +32,195 @@ import (
...
+	go s.startReceivingBlocks()
 
-	for {
-		msg, ok := <-s.msgRec
-		if !ok {
-			log.Warn("core service message watcher", "error", "channel closed")
-			break
-		}
+	// start receiving messages from p2p service
+	go s.startReceivingMessages()
```
Comment (ansermino) [inline core/service.go]: I think `start` is a bit of a misnomer. These functions **are** receiving blocks and messages, you start them by putting them in their own routine.

File: core/service_test.go [hunk 7]
```diff
@@ -36,7 +36,10 @@ import (
 const POLKADOT_RUNTIME_FP string = "../substrate_test_runtime.compact.wasm"
 const POLKADOT_RUNTIME_URL string = "https://github.com/noot/substrate/blob/add-blob/core/test-runtime/wasm/wasm32-unknown-unknown/release/wbuild/substrate-test-runtime/substrate_test_runtime.compact.wasm?raw=true"
 
-// getRuntimeBlob checks if the polkadot runtime wasm file exists and if not, it fetches it from github
+var TestStatusTimeout = 2 * time.Second
```
Comment (ansermino) [inline core/service_test.go]: Used in multiple tests not just status

Comment (ryanchristo) [reply core/service_test.go]: True. Will fix.

File: core/service.go (hunk 6 above)
Comment (ryanchristo) [reply core/service.go]: Ah, ok. Maybe just `receiveMessages` then.

File: core/service.go (hunk 5 above)
Comment (ryanchristo) [reply core/service.go]: Ok, will put it all in the return line.

File: consensus/babe/babe.go (hunk 4 above)
Comment (ryanchristo) [reply consensus/babe/babe.go]: This should be a pointer still. Not sure why I did that. Will fix.

**Comment by ryanchristo** [comment]
Will rebase with fixes after #446 is merged.

File: cmd/gossamer/configcmd.go [hunk 8]
```diff
@@ -83,19 +84,20 @@ func makeNode(ctx *cli.Context) (*dot.Dot, *cfg.Config, error) {
 
 	log.Info("🕸\t Configuring node...", "datadir", fig.Global.DataDir, "protocolID", string(gendata.ProtocolId), "bootnodes", fig.P2p.BootstrapNodes)
 
-	// TODO: BABE
-	msgRec := make(chan p2p.Message)
+	bsChan := make(chan types.Block)
```
Comment (ryanchristo) [inline cmd/gossamer/configcmd.go]: Not sure what to name the channel before it becomes a send or receive channel...

Comment (ansermino) [reply cmd/gossamer/configcmd.go]: See comment below

File: cmd/gossamer/configcmd.go [hunk 9]
```diff
@@ -83,19 +84,20 @@ func makeNode(ctx *cli.Context) (*dot.Dot, *cfg.Config, error) {
...
+	p2pSrvc, p2pSend := createP2PService(fig, gendata)
 	srvcs = append(srvcs, p2pSrvc)
 
 	// core.Service
 	coreCfg := &core.ServiceConfig{
 		Keystore: ks,
 		Runtime:  r,
-		MsgRec:   msgRec,
-		MsgSend:  msgSend,
+		BsChan:   bsChan,
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply cmd/gossamer/configcmd.go]: fixed

File: consensus/babe/babe.go [hunk 10]
```diff
@@ -28,42 +28,36 @@ import (
...
 	authorityData  []AuthorityData
-
 	epochThreshold *big.Int // validator threshold for this epoch
 	txQueue        *tx.PriorityQueue
-	isProducer     map[uint64]bool // whether we are a block producer at a slot
-
-	// Block announce channel used every time a block is created
-	blockAnnounce chan<- p2p.Message
+	isProducer     map[uint64]bool    // whether we are a block producer at a slot
+	blockSend      chan<- types.Block // send blocks to core service
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply consensus/babe/babe.go]: noew `newBlocks` for babe session

File: cmd/gossamer/configcmd.go [hunk 11]
```diff
@@ -83,19 +84,20 @@ func makeNode(ctx *cli.Context) (*dot.Dot, *cfg.Config, error) {
...
+		P2pRec:   p2pRec,
+		P2pSend:  p2pSend,
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply cmd/gossamer/configcmd.go]: updated to `core.Config`

**Pull request merged**